src/rpc/ngHandlers/Unsubscribe.h

```diff
@@ -0,0 +1,212 @@
+
+public:
+    struct OrderBook
+    {
+        ripple::Book book;
+        bool both = false;
+    };
+
+    struct Input
+    {
+        std::optional<std::vector<std::string>> accounts;
+        std::optional<std::vector<std::string>> streams;
+        std::optional<std::vector<std::string>> accountsProposed;
+        std::optional<std::vector<OrderBook>> books;
+    };
+
+    using Output = VoidOutput;
+    using Result = RPCng::HandlerReturnType<Output>;
+
+    BaseUnsubscribeHandler(
+        std::shared_ptr<BackendInterface> const& sharedPtrBackend,
+        std::shared_ptr<SubscriptionManagerType> const& subscriptions)
+        : sharedPtrBackend_(sharedPtrBackend), subscriptions_(subscriptions)
+    {
+    }
+
+    RpcSpecConstRef
+    spec() const
+    {
+        static auto const booksValidator =
+            validation::CustomValidator{[](boost::json::value const& value, std::string_view key) -> MaybeError {
+                if (!value.is_array())
+                    return Error{RPC::Status{RPC::RippledError::rpcINVALID_PARAMS, std::string(key) + "NotArray"}};
+                for (auto const& book : value.as_array())
+                {
+                    if (!book.is_object())
+                        return Error{
+                            RPC::Status{RPC::RippledError::rpcINVALID_PARAMS, std::string(key) + "ItemNotObject"}};
+                    if (book.as_object().contains("both") && !book.as_object().at("both").is_bool())
+                    {
+                        return Error{RPC::Status{RPC::RippledError::rpcINVALID_PARAMS, "bothNotBool"}};
+                    }
+                    auto const parsedBook = RPC::parseBook(book.as_object());
+                    if (auto const status = std::get_if<RPC::Status>(&parsedBook))
+                        return Error(*status);
+                }
+                return MaybeError{};
+            }};
+
+        static auto const rpcSpec = RpcSpec{
+            {JS(streams), validation::SubscribeStreamValidator},
+            {JS(accounts), validation::SubscribeAccountsValidator},
+            {JS(accounts_proposed), validation::SubscribeAccountsValidator},
+            {JS(books), booksValidator}};
+
+        return rpcSpec;
+    }
+
+    Result
+    process(Input input, Context const& ctx) const
+    {
+        if (input.streams)
+        {
+            unsubscribeToStreams(*(input.streams), ctx.session);
+        }
+        if (input.accounts)
+        {
+            unsubscribeToAccounts(*(input.accounts), ctx.session);
+        }
+        if (input.accountsProposed)
+        {
+            unsubscribeToProposedAccounts(*(input.accountsProposed), ctx.session);
+        }
+        if (input.books)
+        {
+            unsubscribeToBooks(*(input.books), ctx.session);
+        }
+        return Output{};
+    }
+
+private:
+    void
+    unsubscribeToStreams(std::vector<std::string> const& streams, std::shared_ptr<WsBase> const& session) const
```

Comment:
it's probably old code that you copied but consider renaming to something like 'unsubscribeFromStreams' :)

Reply:
yea, 'from' makes more sense